internal/util/basesubjectset.go

```diff
@@ -224,32 +244,581 @@ func (bss BaseSubjectSet[T]) UnionWithSet(other BaseSubjectSet[T]) {
+	// are found in the excluded list. If the wildcard *itself* is conditional, then instead of
+	// items being removed, they are made conditional on the inversion of the wildcard's expression,
+	// and the exclusion's conditional, if any.
+	//
+	// Examples:
+	//  {user:sarah, user:tom} - {*} => {}
+	//  {user:sarah, user:tom} - {*[somecaveat]} => {user:sarah[!somecaveat], user:tom[!somecaveat]}
+	//  {user:sarah, user:tom} - {* - {user:tom}} => {user:tom}
+	//  {user:sarah, user:tom} - {*[somecaveat] - {user:tom}} => {user:sarah[!somecaveat], user:tom}
+	//  {user:sarah, user:tom} - {* - {user:tom[c2]}}[somecaveat] => {user:sarah[!somecaveat], user:tom[c2]}
+	//  {user:sarah[c1], user:tom} - {*[somecaveat] - {user:tom}} => {user:sarah[c1 && !somecaveat], user:tom}
+	exclusions := exclusionsMapFor(wildcard)
+	for existingSubjectID, existingSubject := range cst.subjects {
+		exclusion, isExcluded := exclusions[existingSubjectID]
+		if !isExcluded {
+			// If the subject was not excluded within the wildcard, it is either removed directly
+			// (in the case where the wildcard is not conditional), or has its condition updated to
+			// reflect that it is only present when the condition for the wildcard is *false*.
+			if wildcard.GetConditionalExpression() == nil {
+				delete(cst.subjects, existingSubjectID)
+				continue
+			}
+
+			cst.subjects[existingSubjectID] = cst.constructor(
+				existingSubjectID,
+				and(existingSubject.GetConditionalExpression(), invert(wildcard.GetConditionalExpression())),
+				nil,
+				existingSubject)
+			continue
+		}
+
+		// If the exclusion is not conditional, then the subject is always present.
+		if exclusion.GetConditionalExpression() == nil {
+			continue
+		}
+
+		// The conditional of the exclusion is that of the exclusion itself OR the inverted case of
+		// the wildcard, which would mean the wildcard itself does not apply.
+		exclusionConditional := union(invert(wildcard.GetConditionalExpression()), exclusion.GetConditionalExpression())
+
+		// If the whole exclusion is not conditional, nothing more to do.
+		if exclusionConditional == nil {
+			continue
+		}
+
+		// Otherwise, add the exclusion's conditional to the entry.
+		cst.subjects[existingSubjectID] = cst.constructor(
+			existingSubjectID,
+			and(existingSubject.GetConditionalExpression(), exclusionConditional),
+			nil,
+			existingSubject)
+	}
+}
+
+func (cst *concreteSubjectSet[T]) subtractConcrete(concrete T) {
+	existing, ok := cst.subjects[concrete.GetSubjectId()]
+	if !ok {
+		// If not present, nothing more to do.
+		return
+	}
+
+	// Subtraction of a concrete type removes the entry from the concrete list
+	// *unless* the subtraction is conditional, in which case the conditional is updated
+	// to remove the element when it is true.
+	//
+	// Examples:
+	//  {user:sarah} - {user:tom} => {user:sarah}
+	//  {user:tom} - {user:tom} => {}
+	//  {user:tom[c1]} - {user:tom} => {user:tom[c1]}
```

Comment:
this is surprising

I would read this as: `maybe tom - definitely tom = maybe tom` when intuitively I would think it should be `maybe tom - definitely tom = {}`

Reply:
Good point... that's definitely not right. I'll fix it